presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java

```diff
@@ -42,16 +42,39 @@ private WorkProcessorUtils() {}
     {
         return new AbstractIterator<T>()
         {
+            final Iterator<Optional<T>> yieldingIterator = yieldingIteratorFrom(processor);
+
             @Override
             protected T computeNext()
+            {
+                while (true) {
+                    if (!yieldingIterator.hasNext()) {
+                        return endOfData();
+                    }
+
+                    Optional<T> elementOptional = yieldingIterator.next();
+                    if (elementOptional.isPresent()) {
+                        return elementOptional.get();
+                    }
+                }
+            }
+        };
+    }
+
+    static <T> Iterator<Optional<T>> yieldingIteratorFrom(WorkProcessor<T> processor)
+    {
+        return new AbstractIterator<Optional<T>>()
+        {
+            @Override
+            protected Optional<T> computeNext()
             {
                 while (true) {
                     if (processor.process()) {
                         if (processor.isFinished()) {
                             return endOfData();
                         }
 
-                        return processor.getResult();
+                        return Optional.of(processor.getResult());
                     }
                     else if (processor.isBlocked()) {
                         try {
```

Comment:
here we should throw: blocking work processor cannot be (safely) iterated over (add this as a first commit)

then you don't need the `while(true)` in `yieldingIteratorFrom()`

Reply:
addressed in #10768

Reply:
Thanks. Can you rebase now?

Reply:
done